src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -1396,13 +1407,25 @@ trait Infer extends Checkable {
           case tp               => tp
         }
 
+        // if at least one retry is required, and the overload includes a Java-defined alternative
+        // proceed without tupling for applicability (if flag is enabled)
+        private[this] var retries = 0
+        private def atRetryDepth[A](cond: Boolean)(body: => A): A = {
+          val res =
+            if (cond && retries > 0) withoutTuplingForApplicability(body)
+            else body
+          retries += 1
+          res
+        }
```

Comment:
I don't really follow this mechanism. What is being retried? Is this related to `isLastTry`?